contracts/protocol/PositionManager.sol

```diff
@@ -604,27 +604,35 @@ contract PositionManager is
         uint8 isFullBuy = 0;
         bool isSkipFirstPip;
         uint256 passedPipCount = 0;
-        CurrentLiquiditySide currentLiquiditySide = CurrentLiquiditySide(
-            _initialSingleSlot.isFullBuy
-        );
-        if (currentLiquiditySide != CurrentLiquiditySide.NotSet) {
-            if (_isBuy)
+        {
+            CurrentLiquiditySide currentLiquiditySide = CurrentLiquiditySide(
+                _initialSingleSlot.isFullBuy
+            );
+            if (currentLiquiditySide != CurrentLiquiditySide.NotSet) {
+                if (_isBuy)
                 // if buy and latest liquidity is buy. skip current pip
-                isSkipFirstPip =
+                    isSkipFirstPip =
                     currentLiquiditySide == CurrentLiquiditySide.Buy;
                 // if sell and latest liquidity is sell. skip current pip
-            else
-                isSkipFirstPip =
+                else
+                    isSkipFirstPip =
                     currentLiquiditySide == CurrentLiquiditySide.Sell;
+            }
         }
-        while (state.remainingSize != 0) {
+        bool onlyLoopOnce;
+        while (!onlyLoopOnce && state.remainingSize != 0) {
             StepComputations memory step;
             // updated findHasLiquidityInMultipleWords, save more gas
-            (step.pipNext) = liquidityBitmap.findHasLiquidityInMultipleWords(
-                state.pip,
-                maxFindingWordsIndex,
-                !_isBuy
-            );
+            if (_maxPip != 0) {
+                step.pipNext = _maxPip;
+                onlyLoopOnce = true;
```

Comment:
only loop once when it's a market fill in a limit order at the same pip. we don't need to loop and find liquidity in all words anymore cause it's just a market at the current pip. saving a lot of gas